File: frontend/src/api/trust.ts
import client from './client';

export type PolicyStatus = 'DRAFT' | 'ACTIVE' | 'REVOKED';
export type CertState = 'PENDING' | 'PASSED' | 'FAILED' | 'REVOKED';
export type TrustLevel = 'certified' | 'partial' | 'unverified';

export interface Policy {
  policy_id: string;
  name: string;
  purpose: string | null;
  description: string | null;
  template: Record<string, unknown>;
  status: PolicyStatus;
  created_at: string;
  updated_at: string;
  applied_count: number;
}

export interface Certification {
  certification_id: string;
  name: string;
  certifier_id: string | null;
  state: CertState;
  notes: string | null;
  issued_at: string | null;
  expires_at: string | null;
  created_at: string;
  linked_count: number;
}

export interface TrustBadge {
  id: string;
  name: string;
  state: string;
}

export interface AgentTrust {
  agent_id: string;
  name: string;
  version: string;
  status: string;
  visibility: string;
  owner_name: string | null;
  certifications: TrustBadge[];
  policies: TrustBadge[];
  trust_level: TrustLevel;
}

export interface TrustOverview {
  total_agents: number;
  certified_agents: number;
  partial_agents: number;
  unverified_agents: number;
  pending_certifications: number;
  active_policies: number;
  draft_policies: number;
  uncertified_exposed_agents: number;
}

export const trustApi = {
  getOverview: () => client.get<TrustOverview>('/trust/overview').then((r) => r.data),
  getAgents: () => client.get<AgentTrust[]>('/trust/agents').then((r) => r.data),

  getPolicies: () => client.get<Policy[]>('/trust/policies').then((r) => r.data),
  createPolicy: (payload: {
    name: string;
    purpose?: string;
    description?: string;
    template?: Record<string, unknown>;
    status?: PolicyStatus;
  }) => client.post<Policy>('/trust/policies', payload).then((r) => r.data),
  updatePolicyStatus: (policyId: string, status: PolicyStatus) =>
    client
      .patch<Policy>(`/trust/policies/${policyId}/status`, { status })
      .then((r) => r.data),

  getCertifications: () =>
    client.get<Certification[]>('/trust/certifications').then((r) => r.data),
  createCertification: (payload: {
    name: string;
    notes?: string;
    state?: CertState;
    expires_at?: string | null;
  }) =>
    client.post<Certification>('/trust/certifications', payload).then((r) => r.data),
  updateCertState: (certId: string, state: CertState, notes?: string) =>
    client
      .patch<Certification>(`/trust/certifications/${certId}/state`, { state, notes })
      .then((r) => r.data),

  linkPolicy: (agentId: string, policyId: string) =>
    client.post(`/trust/agents/${agentId}/policies`, { policy_id: policyId }),
  unlinkPolicy: (agentId: string, policyId: string) =>
    client.delete(`/trust/agents/${agentId}/policies/${policyId}`),
  linkCertification: (agentId: string, certificationId: string) =>
    client.post(`/trust/agents/${agentId}/certifications`, {
      certification_id: certificationId,
    }),
  unlinkCertification: (agentId: string, certId: string) =>
    client.delete(`/trust/agents/${agentId}/certifications/${certId}`),
};

File: backend/app/schemas/trust.py
"""MeshBoard — 신뢰 관리(Trust Workbench) 스키마.

정책(Policy)·인증(Certification) 관리와 에이전트 신뢰 현황 응답 모델.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

POLICY_STATUSES = {"DRAFT", "ACTIVE", "REVOKED"}
CERT_STATES = {"PENDING", "PASSED", "FAILED", "REVOKED"}


# ── Policy ────────────────────────────────────────────────────────
class PolicyCreate(BaseModel):
    name: str = Field(..., max_length=255)
    purpose: Optional[str] = None
    description: Optional[str] = None
    template: Dict[str, Any] = Field(default_factory=dict)
    status: str = Field(default="DRAFT", max_length=20)


class PolicyStatusUpdate(BaseModel):
    status: str = Field(..., max_length=20)


class PolicyTemplateValidate(BaseModel):
    template: Dict[str, Any] = Field(default_factory=dict)


class PolicyTemplateValidationResult(BaseModel):
    valid: bool
    errors: List[str] = Field(default_factory=list)
    supported_fields: List[str] = Field(default_factory=list)


class PolicyRead(BaseModel):
    policy_id: uuid.UUID
    name: str
    purpose: Optional[str] = None
    description: Optional[str] = None
    template: Dict[str, Any] = Field(default_factory=dict)
    status: str
    created_at: datetime
    updated_at: datetime
    applied_count: int = 0

    model_config = {"from_attributes": True}


# ── Certification ─────────────────────────────────────────────────
class CertificationCreate(BaseModel):
    name: str = Field(..., max_length=255)
    notes: Optional[str] = None
    state: str = Field(default="PENDING", max_length=20)
    expires_at: Optional[datetime] = None


class CertificationStateUpdate(BaseModel):
    state: str = Field(..., max_length=20)
    notes: Optional[str] = None


class CertificationRead(BaseModel):
    certification_id: uuid.UUID
    name: str
    certifier_id: Optional[uuid.UUID] = None
    state: str
    notes: Optional[str] = None
    issued_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    created_at: datetime
    linked_count: int = 0

    model_config = {"from_attributes": True}


# ── Agent Trust Posture ───────────────────────────────────────────
class TrustBadge(BaseModel):
    id: uuid.UUID
    name: str
    state: str  # cert state 또는 policy status


class AgentTrustRead(BaseModel):
    agent_id: uuid.UUID
    name: str
    version: str
    status: str
    visibility: str
    owner_name: Optional[str] = None
    certifications: List[TrustBadge] = Field(default_factory=list)
    policies: List[TrustBadge] = Field(default_factory=list)
    trust_level: str  # "certified" | "partial" | "unverified"


# ── Linking ───────────────────────────────────────────────────────
class AgentPolicyLink(BaseModel):
    policy_id: uuid.UUID


class AgentCertLink(BaseModel):
    certification_id: uuid.UUID


# ── Overview ──────────────────────────────────────────────────────
class TrustOverview(BaseModel):
    total_agents: int
    certified_agents: int
    partial_agents: int
    unverified_agents: int
    pending_certifications: int
    active_policies: int
    draft_policies: int
    uncertified_exposed_agents: int  # PUBLIC/DEPARTMENT 인데 인증 없는 에이전트
